icubam/predicu/plot/__init__.py

```diff
@@ -116,28 +120,57 @@ def plot(
   matplotlib_style: str,
   restrict_to_region: Optional[str] = None,
 ):
+  """Generate one plot
+
+  Args:
+    plot_name: name of the plot to make. Must match one of the files in plot/
+    cached_data : a dictionary with **raw** data for different sources.
+  """
   plot_module = __import__(f"{plot_name}", globals(), locals(), ["plot"], 1)
   plot_fun = plot_module.plot  # type: ignore
-  data_source = plot_module.data_source  # type: ignore
+
+  data_source = plot_module.data_source.copy()  # type: ignore
+  needs_combined_data = 'combined_bedcounts_public' in data_source
+  if needs_combined_data:
+    data_source.remove("combined_bedcounts_public")
+    # load requied data to make combined dataset
+    data_source += ['bedcounts', 'public']
+
+  # load required raw data if it's missing
   for name in data_source:
-    cached_data[name] = load_if_not_cached(
-      name,
-      cached_data,
-      api_key=api_key,
-      icubam_host=icubam_host,
-      restrict_to_region=restrict_to_region,
+    if name not in cached_data:
+      cached_data[name] = load_data(
+        name,
+        api_key=api_key,
+        icubam_host=icubam_host,
+      )
+
+  # preprocess the subset of data necessary this plot
+  plot_data = {}
+  for name in data_source:
+    data = cached_data[name].copy()
+    if name == "bedcounts":
+      preprocess_kwargs = {'restrict_to_region': restrict_to_region}
+    else:
+      preprocess_kwargs = {}
+
+    plot_data[name] = preprocess_data(name, data, **preprocess_kwargs)
+
+  if needs_combined_data:
+    plot_data["combined_bedcounts_public"] = combine_bedcounts_public(
+      plot_data['public'], plot_data['bedcounts']
     )
+    data_source.remove("bedcounts")
+    data_source.remove("public")
+    data_source.append("combined_bedcounts_public")
```

Comment:
There is quite a bit of logic in this function to fetch the data that is necessary for plots. Not sure if there is a better way.

Reply:
maybe create a fonction `get_raw_sources_needed_for_data_source(data_source : str)` such that
```python
raw_sources = get_raw_sources_needed_for_data_source("combined_bedcounts_public")
assert raw_sources == {"bedcounts", "public"}`
```
maybe that would make the code clearer

Reply:
Thanks a lot for the review! Yes, I'll try to do it in a follow up PR, as this one has already way too many changes :)